jetbrains-core/src/software/aws/toolkits/jetbrains/services/lambda/execution/sam/SamRunningState.kt

```diff
@@ -3,108 +3,238 @@
+    private fun buildLambdaFromTemplate(templateFile: VirtualFile, logicalId: String, samOptions: SamOptions): BuildRequest {
+        val templatePath = Paths.get(templateFile.path)
+        val buildDir = templatePath.resolveSibling(".aws-sam").resolve("build")
 
-        // Unix: Sends SIGINT on destroy so Docker container is shut down
-        // Windows: Run with mediator to allow for Cntrl+C to be used
-        return object : KillableColoredProcessHandler(commandLine, true) {
-            override fun doDestroyProcess() {
-                // send signal only if user explicitly requests termination
-                if (this.getUserData(ProcessHandler.TERMINATION_REQUESTED) == true) {
-                    super.doDestroyProcess()
+        return BuildRequest(templatePath, logicalId, emptyMap(), buildDir)
+    }
+
+    private fun buildLambdaFromHandler(lambdaBuilder: LambdaBuilder, project: Project, settings: HandlerRunSettings): BuildRequest {
+        val samOptions = settings.samOptions
+        val runtime = settings.runtime
+        val handler = settings.handler
+
+        val element = Lambda.findPsiElementsForHandler(project, runtime, handler).first()
+        val module = getModule(element.containingFile)
+
+        val buildDirectory = lambdaBuilder.getBuildDirectory(module)
+        val dummyTemplate = buildDirectory.parent.resolve("temp-template.yaml")
+        val dummyLogicalId = "Function"
+
+        SamTemplateUtils.writeDummySamTemplate(
+            tempFile = dummyTemplate,
+            logicalId = dummyLogicalId,
+            runtime = runtime,
+            handler = handler,
+            timeout = settings.timeout,
+            memorySize = settings.memorySize,
+            codeUri = lambdaBuilder.handlerBaseDirectory(module, element).toAbsolutePath().toString(),
+            envVars = settings.environmentVariables
+        )
+
+        return BuildRequest(
+            dummyTemplate,
+            dummyLogicalId,
+            lambdaBuilder.additionalBuildEnvironmentVariables(module, samOptions),
+            buildDirectory
+        )
+    }
+
+    private fun buildWorkflow(
+        environment: ExecutionEnvironment,
+        settings: LocalLambdaRunSettings,
+        state: SamRunningState,
+        buildRequest: BuildRequest,
+        emitter: BuildView
+    ): ProcessHandler {
+        val buildStep = BuildLambda(buildRequest.template, buildRequest.logicalId, buildRequest.buildDir, buildRequest.buildEnvVars, settings.samOptions)
+        val startSam = SamRunnerStep(environment, settings, environment.isDebug())
+
+        val workflow = StepWorkflow(
+            buildList {
+                add(ValidateDocker())
+                add(buildStep)
+                if (environment.isDebug()) {
+                    add(GetPorts(settings))
+                    add(object : ParallelStep() {
```

Comment:
is this parallel because the sam step won't complete while it's running?

Reply:
yep we don't have like an "async step" or something like that